src/pybind/mgr/balancer/module.py

```diff
@@ -603,31 +650,118 @@ def do_crush_compat(self, plan):
                          overlap)
             return False
 
-        key = 'pgs'  # pgs objects or bytes
+        # pgs objects bytes or auto
+        '''
+        In auto-mode, alternate between objects, bytes and pgs.
+        The relative importance between PGs, objects and bytes is out of the fact that PGs have much
+        more role in the distribution in the long run than the others. The parameter 'bytes' varies
+        rather too rapidly, 'objects' would vary less frequently and PGs the slowest. The rate of varying
+        of these parameters is inversely proportional to how much they affect the distribution in the long run.
+        '''
+        key = self.get_config('key', default_key)
+        max_iterations = self.max_iterations
+        no_improvement = 0
+        improve_tolerance = 10
+        previous_score = None
+        tolerance_threshold = 0.000001
+        best_weights = {}
+
+        key_copy = self.get_config('key', default_key)
+        if(key == 'auto'):
+            key = 'pgs'
 
         # go
-        random.shuffle(roots)
-        for root in roots:
-            pools = pe.root_pools[root]
-            self.log.info('Balancing root %s (pools %s) by %s' %
-                          (root, pools, key))
-            target = pe.target_by_root[root]
-            actual = pe.actual_by_root[root][key]
-            queue = sorted(actual.keys(),
-                           key=lambda osd: -abs(target[osd] - actual[osd]))
-            self.log.debug('queue %s' % queue)
-            for osd in queue:
-                deviation = target[osd] - actual[osd]
-                if deviation == 0:
+        for iterations in range(max_iterations):
```

Comment:
In case of py2, `range()` will actually return a list instead of a generator. So, the only problem will be that we will need a bit more space in memory.
But if we keep it this way, the code will be support both py2 and py3, instead of having to write separate codes for py2 and py3 and also the additional need of importing `sys` module to check the python version.